src/expr/src/scalar/func.rs

```diff
@@ -1166,6 +1166,82 @@ fn add_interval<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError>
         .map(Datum::from)
 }
 
+fn bit_not_int16<'a>(a: Datum<'a>) -> Datum<'a> {
+    Datum::from(!a.unwrap_int16())
+}
+
+fn bit_not_int32<'a>(a: Datum<'a>) -> Datum<'a> {
+    Datum::from(!a.unwrap_int32())
+}
+
+fn bit_not_int64<'a>(a: Datum<'a>) -> Datum<'a> {
+    Datum::from(!a.unwrap_int64())
+}
+
+fn bit_and_int16<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int16() & b.unwrap_int16())
+}
+
+fn bit_and_int32<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int32() & b.unwrap_int32())
+}
+
+fn bit_and_int64<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int64() & b.unwrap_int64())
+}
+
+fn bit_or_int16<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int16() | b.unwrap_int16())
+}
+
+fn bit_or_int32<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int32() | b.unwrap_int32())
+}
+
+fn bit_or_int64<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int64() | b.unwrap_int64())
+}
+
+fn bit_xor_int16<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int16() ^ b.unwrap_int16())
+}
+
+fn bit_xor_int32<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int32() ^ b.unwrap_int32())
+}
+
+fn bit_xor_int64<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int64() ^ b.unwrap_int64())
+}
+
+fn bit_shift_left_int16<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    // widen to i32 and then cast back to i16 in order to align with the behavior of Postgres
+    // when the number of bits is in the 16-31 range
+    Datum::from(((a.unwrap_int16() as i32) << (b.unwrap_int32() & 0b_0001_1111)) as i16)
+}
+
+fn bit_shift_left_int32<'a>(a: Datum<'a>, b: Datum<'a>) -> Datum<'a> {
+    Datum::from(a.unwrap_int32() << (b.unwrap_int32() & 0b_0001_1111))
```

Comment:
Rust will panic if `b` is negative but postgres has overflow semantics. You can use [`overflow_shl`](https://doc.rust-lang.org/std/primitive.i32.html#method.overflowing_shl) and friends in all these functions to have the same behaviour.

Also, it took me a while to figure out what's going on with the `& 0b_0001_1111` part and from some quick tests it doesn't seem necessary to me. This produces the same results as postgres for me:

```rust
    a.unwrap_i32().overflowing_shl(b.unwrap_i32() as u32).0
```

Reply:
I noticed today that the bitmask also handles negatives so that's not an issue with the current code but I still think we should use `overflowing_shl` since that's the behaviour we're after

Reply:
> Rust will panic if b is negative but postgres has overflow semantics

I am covering negative `b` values in the tsts ([here](https://github.com/aalexandrov/materialize/blob/sql_bitwise_operators/test/sqllogictest/arithmetic.slt#L589) and [here](https://github.com/aalexandrov/materialize/blob/sql_bitwise_operators/test/sqllogictest/arithmetic.slt#L589)) and they work as expected (result tables are taken from Postgres).

Will take a look at the `overflowing_shl` & friends.